### ISSUE TITLE
Disccusion on Exploration and Design of Interactive Elements in Spoke

### ISSUE THREAD

**Issue by misslivirose** [mozilla/Spoke#857]
**Background**
We have received a fair amount of feedback that Spoke could support some degree of interactive components in published scenes. Examples of types of interactivity that have been requested include: 

- Basic transform-driven animations (e.g. rotate Object A around Object B, or move Object C along a specific path) 
- Creating a state-driven narration system to guide someone through a scene 
- Playing sound effects from different objects in the scene when they are approached
- Changing colors, meshes, or other properties on a node when some specified event occurs

**Design Considerations**
Our discussions for this feature have centered largely around the ability to provide a visual scripting style interface in Spoke. @robertlong has done a design pass and background research into node-based editors, including: 
- layout flow
- variable scoping 
- various levels of abstraction for concepts

**Existing Work in Spoke**
Spoke currently has an experimental "Trigger Volume" feature that can be enabled that creates an element that has a very small subset of interactive properties. The trigger volume can play or pause a video node that is in the scene, or be used to play or stop an animation that is included in a glTF model. These experimental features are not intended for production-level scene publication, and may not be backwards-compatible with future work done to allow for the creation of interactive elements. 

**First release exclusions**
- Custom shaders for meshes or surfaces
- Ability to write arbitrary JavaScript

**Relation to custom components work**
In parallel with any work that we do to facilitate interactive elements in Spoke, we should also consider relevant work that is or could be done to more widely facilitate the creation of custom components for Hubs. However, it can be anticipated that the audiences for each of these projects will be different.

**Comment by EgleXR** [comment]
I suggest adding the functionality of touching an object, transporting the user to a previously defined way point.

**Comment by BluePrintRandom** [comment]
I would suggest SCA logic bricks, or logic nodes to make the behaviors the most flexible long term.

and defining multiple 'triggers' that execute this code. as well as permission levels you can give an object.

**Comment by blairmacintyre** [comment]
I would be careful on the scope because this sort of thing has the potential to become a massive resource sucking monster.  Understanding what is really useful / needed seems hard.  Sure, we'd like to be able to completely program an interactive experience, but without creating a full-blown programming environment, what are the problems that can be solved relative to the things that could be done _instead_?

For example, with the ability to self-host, what percentage of people who want to leverage this also want to self-host, and would be better served by the ability to write and use custom components in their hosted client?

**Comment by blairmacintyre** [comment]
In particular, what is the balance between this an #1976, and where does programmer-oriented custom code come in?  People (i.e, me) will want to do more than either this or #1976 implies.  For example, I will likely want to create custom content elements akin to AFrame components that let me do some three.js content completely under my own control.

I fully understand that such components are likely to break, especially if a future internal architecture does something radical like move from AFrame.  But one of the benefits of custom hosting is that I _don't need to upgrade_ ... if I make a custom setup and need time to update it when hubs changes, I can take my time and upgrade when things are ready.

**Comment by ocularvr** [comment]
For us, just being able to start/stop video and glb animations would just go a real long way. I think we all know that in 10 years Spoke will look like Unreal or Unity but today, we are grappling with stones :) 
If we could send an argument in the chat box to start and stop these two media types we will get to bronze age by teleporting ;)

**Comment by nathank000** [comment]
All - checking in on this from the future,
I am running Hubs on a personal AWS account and am evaluating the project.
Wondering if interactivity (shared state, IO, etc) is possible/ has been built out?
